core/genesis/file.go

```diff
@@ -0,0 +1,261 @@
+	fName := fmt.Sprintf("%s.json", field)
+	fPath := filepath.Join(f.sourceDir, f.moduleName)
+
+	fp, err := os.Open(filepath.Clean(filepath.Join(fPath, fName)))
+	if err == nil {
+		return fp, nil
+	}
+
+	if !os.IsNotExist(err) {
+		return nil, fmt.Errorf("unexpected error: %w", err)
+	}
+
+	// if cannot find it, try reading from <sourceDir>/<module>.json
+	rawBz, err := f.ReadRawJSON()
+	if err == nil {
+		f.moduleRootJson = rawBz
+		return f.unmarshalRawModuleWithField(rawBz, field)
+	}
+
+	if !os.IsNotExist(err) {
+		return nil, err
+	}
+
+	if f.moduleRootJson == nil {
+		return nil, fmt.Errorf("failed to retrieve the state of %s from moduleRootJson", f.moduleName)
+	}
+	return f.unmarshalRawModuleWithField(f.moduleRootJson, field)
+}
+
+func (f *FileGenesisSource) unmarshalRawModuleWithField(rawBz []byte, field string) (io.ReadCloser, error) {
+	fieldState := make(map[string]json.RawMessage)
+	err := json.Unmarshal(rawBz, &fieldState)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal fields from module state %s, err: %w", f.moduleName, err)
+	}
+
+	fieldRawData := fieldState[field]
+	if fieldRawData == nil {
+		return nil, fmt.Errorf("failed to retrieve module field %s/%s from genesis.json", f.moduleName, field)
+	}
+
+	// wrap raw field data to a ReadCloser
+	return io.NopCloser(bytes.NewReader(fieldRawData)), nil
+}
+
+// ReadMessage reads rawJSON data from source file or moduleRawData,
+// and then unmarshal it into proto.Message
+func (f *FileGenesisSource) ReadMessage(msg proto.Message) error {
+	bz, err := f.ReadRawJSON()
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("unexpected error: %w", err)
+		}
+
+		// read the data from moduleRootJson if the source file doesn't exist
+		bz = f.moduleRootJson
+	}
+	return protojson.Unmarshal(bz, msg)
+}
+
+// ReadRawJSON returns a json.RawMessage read from the source file given by the
+// source directory and the module name.
+func (f *FileGenesisSource) ReadRawJSON() (rawBz json.RawMessage, rerr error) {
+	fName := fmt.Sprintf("%s.json", f.moduleName)
+	fPath := filepath.Join(f.sourceDir, fName)
+
+	fp, err := os.Open(filepath.Clean(fPath))
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err := fp.Close(); err != nil {
+			if rerr != nil {
+				rerr = fmt.Errorf("failed to close file %s: %s, %w", fp.Name(), err.Error(), rerr)
+				return
+			}
+			rerr = fmt.Errorf("failed to close file %s: %w", fp.Name(), err)
+		}
+	}()
+
+	fi, err := fp.Stat()
+	if err != nil {
+		rerr = fmt.Errorf("failed to stat file %s: %w", fp.Name(), rerr)
+		return nil, rerr
+	}
+
+	buf, err := io.ReadAll(fp)
+	if err != nil {
+		rerr = fmt.Errorf("failed to read file %s: %w", fp.Name(), err)
+		return nil, rerr
+	}
+	if int64(len(buf)) != fi.Size() {
+		rerr = fmt.Errorf("couldn't read entire file: %s, read: %d, file size: %d", fp.Name(), len(buf), fi.Size())
+		return nil, rerr
+	}
+	return buf, nil
+}
```

Comment:
This should default to the json coming from initchain

Reply:
Hmm, I updated it to return the default rawJson when the file doesn't exist. If we have an error during the file reading, we should return the error directly I think.

Reply:
Makes sense